File: backupx.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <ctime>
#include <iomanip>

using namespace std;

// Banner
const string BANNER = R"(
   ______     ______     ______     ______     ______     ______     ______     ______     ______  
  /\  == \   /\  __ \   /\  ___\   /\  __ \   /\  == \   /\  ___\   /\  == \   /\  ___\   /\__  _\ 
  \ \  __<   \ \  __ \  \ \ \____  \ \  __ \  \ \  __<   \ \ \____  \ \  __<   \ \  __\   \/_/\ \/ 
   \ \_\ \_\  \ \_\ \_\  \ \_____\  \ \_\ \_\  \ \_____\  \ \_____\  \ \_\ \_\  \ \_____\    \ \_\ 
    \/_/ /_/   \/_/\/_/   \/_____/   \/_/\/_/   \/_____/   \/_____/   \/_/ /_/   \/_____/     \/_/ 
                                                                                                   
)";

void mainProgram(int argc, char* argv[]) {
    // Argument parsing
    vector<string> extensions;
    string extensionFile, inputListFile, inputFileName, date, numberSeparator, dateSeparator, dateFormat;
    int numberRange = 0;
    bool silent = false;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "-e" && i + 1 < argc) extensionFile = argv[++i];
        else if (arg == "-l" && i + 1 < argc) inputListFile = argv[++i];
        else if (arg == "-i" && i + 1 < argc) inputFileName = argv[++i];
        else if (arg == "-d" && i + 1 < argc) date = argv[++i];
        else if (arg == "-n" && i + 1 < argc) numberRange = stoi(argv[++i]);
        else if (arg == "-ns" && i + 1 < argc) numberSeparator = argv[++i];
        else if (arg == "-ds" && i + 1 < argc) dateSeparator = argv[++i];
        else if (arg == "-df" && i + 1 < argc) dateFormat = argv[++i];
        else if (arg == "-silent") silent = true;
    }

    if (!silent) cout << BANNER << endl;

    // Read extensions from text file
    ifstream extensionFileStream(extensionFile);
    if (!extensionFileStream.is_open()) {
        cout << "Extension file not found." << endl;
        return;
    }
    string line;
    while (getline(extensionFileStream, line)) {
        extensions.push_back(line);
    }
    extensionFileStream.close();

    // Prepare user inputs
    vector<string> userInputs;
    if (!inputListFile.empty()) {
        ifstream inputListFileStream(inputListFile);
        if (inputListFileStream.is_open()) {
            while (getline(inputListFileStream, line)) {
                userInputs.push_back(line);
            }
            inputListFileStream.close();
        } else {
            cout << "Input list file not found." << endl;
            return;
        }
    }
    if (!inputFileName.empty()) {
        userInputs.push_back(inputFileName);
    }

    // Add extensions to user input
    vector<string> outputs;
    if (!date.empty()) {
        int year = stoi(date);
        for (int month = 1; month <= 12; ++month) {
            for (int day = 1; day <= 31; ++day) {
                try {
                    tm dateObj = {};
                    dateObj.tm_year = year - 1900;
                    dateObj.tm_mon = month - 1;
                    dateObj.tm_mday = day;
                    mktime(&dateObj);
                    stringstream formattedDate;
                    if (dateFormat == "year") {
                        formattedDate << dateObj.tm_year + 1900;
                    } else {
                        if (!dateSeparator.empty()) formattedDate << dateSeparator;
                        formattedDate << setw(4) << setfill('0') << dateObj.tm_year + 1900;
                        formattedDate << setw(2) << setfill('0') << dateObj.tm_mon + 1;
                        formattedDate << setw(2) << setfill('0') << dateObj.tm_mday;
                    }
                    for (const string& userInput : userInputs) {
                        for (const string& extension : extensions) {
                            outputs.push_back(userInput + "-" + formattedDate.str() + "." + extension);
                        }
                    }
                } catch (const exception&) {
                }
            }
        }
    } else if (numberRange > 0) {
        for (int num = 1; num <= numberRange; ++num) {
            for (const string& userInput : userInputs) {
                for (const string& extension : extensions) {
                    outputs.push_back(userInput + (numberSeparator.empty() ? "" : numberSeparator) + to_string(num) + "." + extension);
                }
            }
        }
    } else {
        for (const string& userInput : userInputs) {
            for (const string& extension : extensions) {
                outputs.push_back(userInput + "." + extension);
            }
        }
    }

    // Print the outputs
    for (const string& output : outputs) {
        cout << output << endl;
    }
}

int main(int argc, char* argv[]) {
    mainProgram(argc, argv);
    return 0;
}

File: backupx.py
import argparse
import json
import datetime

# Banner
BANNER = """
   ______     ______     ______     ______     ______     ______     ______     ______     ______  
  /\  == \   /\  __ \   /\  ___\   /\  __ \   /\  == \   /\  ___\   /\  == \   /\  ___\   /\__  _\ 
  \ \  __<   \ \  __ \  \ \ \____  \ \  __ \  \ \  __<   \ \ \____  \ \  __<   \ \  __\   \/_/\ \/ 
   \ \_\ \_\  \ \_\ \_\  \ \_____\  \ \_\ \_\  \ \_____\  \ \_____\  \ \_\ \_\  \ \_____\    \ \_\ 
    \/_/ /_/   \/_/\/_/   \/_____/   \/_/\/_/   \/_____/   \/_____/   \/_/ /_/   \/_____/     \/_/ 
                                                                                                   

"""

def main():
    parser = argparse.ArgumentParser(description='Add extensions to user input')
    parser.add_argument('-e', '--extension_file', help='Path to text file containing extensions', required=True)
    parser.add_argument('-l', '--input_list_file', help='Path to file containing a list of entries')
    parser.add_argument('-i', '--input_file_name', help='Single file entry')
    parser.add_argument('-d', '--date', help='Year for date-based entries')
    parser.add_argument('-n', '--number_range', type=int, help='Number for range-based entries')
    parser.add_argument('-ns', '--number_separator', help='Separator for attaching numbers')
    parser.add_argument('-ds', '--date_separator', help='Separator for attaching date')
    parser.add_argument('-df', '--date_format', choices=['year', 'ymd'], help='Output format for date-based entries')
    parser.add_argument('-silent', action='store_true', help='Run the program silently without the banner')
    args = parser.parse_args()

    if not args.silent:
        print(BANNER)

    # Read extensions from text file
    try:
        with open(args.extension_file, 'r') as extension_file:
            extensions = extension_file.read().strip().split('\n')
    except FileNotFoundError:
        print("Extension file not found.")
        return

    # Prepare user inputs
    user_inputs = []
    if args.input_list_file:
        try:
            with open(args.input_list_file, 'r') as input_list_file:
                user_inputs = input_list_file.read().strip().split('\n')
        except FileNotFoundError:
            print("Input list file not found.")
            return
    if args.input_file_name:
        user_inputs.append(args.input_file_name)

    # Add extensions to user input
    outputs = []
    if args.date:
        year = int(args.date)
        for month in range(1, 13):
            for day in range(1, 32):
                try:
                    date_obj = datetime.date(year, month, day)
                    if args.date_format == 'year':
                        formatted_date = date_obj.strftime("%Y")
                    else:
                        if args.date_separator:
                            formatted_date = date_obj.strftime("%Y{}%m{}%d".format(args.date_separator, args.date_separator))
                        else:
                            formatted_date = date_obj.strftime("%Y%m%d")
                    outputs.extend([f"{user_input}-{formatted_date}.{extension}" for user_input in user_inputs for extension in extensions])
                except ValueError:
                    pass  # Ignore invalid dates
    elif args.number_range:
        separator = args.number_separator if args.number_separator else ''
        for num in range(1, args.number_range + 1):
            outputs.extend([f"{user_input}{separator}{num}.{extension}" for user_input in user_inputs for extension in extensions])
    else:
        outputs.extend([f"{user_input}.{extension}" for user_input in user_inputs for extension in extensions])

    # Print the outputs
    for output in outputs:
        print(output)

if __name__ == '__main__':
    main()
